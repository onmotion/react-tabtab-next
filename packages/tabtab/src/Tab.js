"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TabStyle = void 0;
var React = __importStar(require("react"));
var styled_components_1 = __importDefault(require("styled-components"));
var CloseButton_1 = __importDefault(require("./CloseButton"));
var TabLi = React.forwardRef(function (props, ref) {
    return React.createElement("li", __assign({ ref: ref }, props), props.children);
});
var TabStyle = (0, styled_components_1.default)(TabLi)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  display: ", ";\n  ", "\n\n  user-select: none;\n  &:hover {\n    cursor: pointer;\n    color: black;\n  }\n"], ["\n  display: ", ";\n  ", "\n\n  user-select: none;\n  &:hover {\n    cursor: pointer;\n    color: black;\n  }\n"])), function (props) { return props.vertical ? 'block' : 'inline-block'; }, function (props) { return props.vertical ?
    "\n      background-color: white;\n      color: black;\n      padding: 10px 10px;\n      z-index: 100000;\n    "
    : function (props) { return props.closable ? 'padding: 10px 10px 10px 15px;' : 'padding: 10px 15px;'; }; });
exports.TabStyle = TabStyle;
var TabText = styled_components_1.default.span(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  vertical-align: middle;\n"], ["\n  vertical-align: middle;\n"])));
var Tab = (function (_super) {
    __extends(Tab, _super);
    function Tab(props) {
        var _this = _super.call(this, props) || this;
        _this.clickTab = _this.clickTab.bind(_this);
        _this.clickDelete = _this.clickDelete.bind(_this);
        return _this;
    }
    Tab.prototype.clickTab = function () {
        var _a = this.props, handleTabChange = _a.handleTabChange, index = _a.index;
        handleTabChange(index);
    };
    Tab.prototype.clickDelete = function (event) {
        event.stopPropagation();
        var _a = this.props, handleEdit = _a.handleEdit, index = _a.index;
        handleEdit({ type: 'delete', index: index });
    };
    Tab.prototype.render = function () {
        var _this = this;
        var _a = this.props, CustomTabStyle = _a.CustomTabStyle, active = _a.active, closable = _a.closable, vertical = _a.vertical, index = _a.index;
        var TabComponent = CustomTabStyle || TabStyle;
        return (React.createElement(TabComponent, { ref: function (node) { return _this.__INTERNAL_NODE = node; }, onClick: this.clickTab, active: active, vertical: vertical, closable: closable, role: "tab", id: "react-tabtab-tab-".concat(index), "aria-controls": "react-tabtab-panel-".concat(index), "aria-selected": active },
            React.createElement(TabText, null, this.props.children),
            closable ?
                React.createElement(CloseButton_1.default, { handleDelete: this.clickDelete })
                : null));
    };
    return Tab;
}(React.PureComponent));
exports.default = Tab;
var templateObject_1, templateObject_2;
