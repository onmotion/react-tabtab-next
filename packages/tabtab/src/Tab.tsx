import * as React from 'react';
import styled from 'styled-components';
import CloseButton from './CloseButton';

export type TabElementProps = React.ComponentPropsWithoutRef<'li'> & TabProps;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const TabElement = React.memo(
    React.forwardRef<HTMLLIElement, TabElementProps>(({ active, closable, vertical, ...props }, ref) => (
        <li ref={ref} {...props}>
            {props.children}
        </li>
    ))
);

const TabStyle = styled(TabElement)`
    display: ${(props) => (props.vertical ? 'block' : 'inline-block')};
    touch-action: auto;
    color: #000000bb;
    border-bottom: 2px solid transparent;
    white-space: nowrap;
    ${(props) =>
        props.vertical
            ? `
      background-color: white;
      color: black;
      padding: 10px 10px;
      z-index: 1;
    `
            : (props) => (props.closable ? 'padding: 10px 10px 8px 15px;' : 'padding: 10px 15px 8px 15px;')}

    user-select: none;
    &:hover,
    &:active {
        cursor: pointer;
        color: black;
    }
    ${(props) =>
        props.active
            ? `
      color: black;
      border-bottom: 2px solid; 
    `
            : null}
`;

const TabText = styled.span`
    vertical-align: middle;
`;

export type TabProps = {
    CustomTabStyle?: React.FC<Partial<TabProps>>;
    handleTabChange?: (event: any) => void;
    handleTabClose?: (event: any) => void;
    index?: number;
    active?: boolean;
    closable?: boolean;
    vertical?: boolean;
    tabIndex?: string;
};

export default class Tab extends React.PureComponent<TabProps> {
    __INTERNAL_NODE: React.ElementRef<any>;

    constructor(props: TabProps) {
        super(props);
        this.clickTab = this.clickTab.bind(this);
        this.clickDelete = this.clickDelete.bind(this);
    }

    clickTab(e: React.MouseEvent<HTMLLIElement, MouseEvent>) {
        e.stopPropagation();
        const { handleTabChange, index } = this.props;
        handleTabChange(index);
    }

    clickDelete(event: React.SyntheticEvent<HTMLButtonElement>) {
        event.stopPropagation(); // prevent trigger clickTab event.
        const { handleTabClose, index } = this.props;
        handleTabClose(index);
    }

    render() {
        const { CustomTabStyle, active, closable, vertical, index } = this.props;
        const TabComponent = CustomTabStyle || TabStyle;

        return (
            <TabComponent
                ref={(node) => (this.__INTERNAL_NODE = node)}
                style={{ touchAction: 'auto' }}
                onClick={this.clickTab}
                active={active}
                vertical={vertical}
                closable={closable}
                role="tab"
                id={`react-tabtab-tab-${index}`}
                aria-controls={`react-tabtab-panel-${index}`}
                aria-selected={active}
            >
                <TabText>{this.props.children}</TabText>
                {closable ? <CloseButton handleTabClose={this.clickDelete} /> : null}
            </TabComponent>
        );
    }
}

export { TabStyle };
