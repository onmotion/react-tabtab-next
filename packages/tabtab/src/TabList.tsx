import * as React from 'react';
import styled from 'styled-components';
import invariant from 'invariant';
import { LeftIcon, RightIcon, BulletIcon } from './IconSvg';
import { isNumber } from './utils/isType';
import { ActionButtonStyle, buttonWidth, ListInner, ListScroll, TabListStyle } from './styledElements';
import { TabListElementProps } from './TabListElement';
import { TabElementProps } from './Tab';
import { PanelProps } from './Panel';
import { SortableContextProps } from '@dnd-kit/sortable';
import { DndContextProps } from '@dnd-kit/core';
import { TabListModal } from './TabListModal';

const makeScrollButton = (ActionButton: React.ElementType) => styled(ActionButton)`
    display: inline-block;
    filter: none;
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    user-select: none;
    ${(props) => (props.left ? (props.showModalButton ? `left: ${buttonWidth + 2}px` : `left: 0`) : 'right: 0')};
    &:hover {
        cursor: pointer;
    }
`;

const makeFoldButton = (ActionButton: React.ElementType) => styled(ActionButton)`
    display: inline-block;
    filter: none;
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    user-select: none;
    left: 0;
    &:hover {
        cursor: pointer;
    }
`;

export type TabListProps = {
    customStyle?: {
        TabList?: React.ElementType<TabListElementProps>;
        Tab?: React.ElementType<TabElementProps>;
        Panel?: React.ElementType<PanelProps>;
        ActionButton?: React.ElementType;
    };
    showArrowButton?: 'auto' | boolean;
    showModalButton?: number | boolean;
    handleTabChange?: (event: any) => void;
    handleTabSequence?: (event: any) => void;
    handleTabClose?: (index: number) => void;
    ExtraButton?: JSX.Element;
    activeIndex?: number;
    children: React.ReactNode[];
    sortableContextProps?: Omit<SortableContextProps, 'children'>;
    dndContextProps?: DndContextProps;
};

type State = {
    modalIsOpen: boolean;
    showArrowButton: 'auto' | boolean;
    showModalButton: boolean | number;
};

export default class TabListComponent extends React.PureComponent<TabListProps, State> {
    listContainer: React.ElementRef<'div'>;
    rightArrowNode: React.ReactElement;
    leftArrowNode: React.ReactElement;
    listScroll: React.ElementRef<'ul'>;
    foldNode: React.ReactElement;
    tabRefs: React.ElementRef<'div'>[];
    scrollPosition: number;
    FoldButton: React.ElementType;
    ScrollButton: React.ElementType;
    TabList: React.ElementType;

    constructor(props: TabListProps) {
        super(props);
        this.handleScroll = this.handleScroll.bind(this);
        this.toggleModal = this.toggleModal.bind(this);
        this.renderTabs = this.renderTabs.bind(this);
        this.renderArrowButtons = this.renderArrowButtons.bind(this);
        this.isShowModalButton = this.isShowModalButton.bind(this);
        this.isShowArrowButton = this.isShowArrowButton.bind(this);
        this.chackActiveIndexRange = this.chackActiveIndexRange.bind(this);
        this.scrollPosition = 0;
        this.tabRefs = [];
        this.TabList = this.props.customStyle?.TabList || TabListStyle;
        this.FoldButton = makeFoldButton(this.props.customStyle?.ActionButton || ActionButtonStyle);
        this.ScrollButton = makeScrollButton(this.props.customStyle?.ActionButton || ActionButtonStyle);

        this.state = {
            modalIsOpen: false,
            showArrowButton: false,
            showModalButton: false,
        };
    }

    chackActiveIndexRange() {
        if (this.props.activeIndex >= this.props.children.length) {
            console.error('activeIndex is out of range 0-' + (this.props.children.length - 1));
            return false;
        }
        return true;
    }

    componentDidMount() {
        console.log('TabList componentDidMount');

        if (!this.chackActiveIndexRange()) return;
        this.isShowArrowButton();
        this.isShowModalButton();

        if (this.props.activeIndex > 0) this.scrollToIndex(this.props.activeIndex, 'left');
    }

    componentDidUpdate(prevProps: TabListProps, prevState: State) {
        if (prevProps.children.length !== this.props.children.length) {
            this.isShowArrowButton();
            this.isShowModalButton();
        }

        if (prevProps.activeIndex !== this.props.activeIndex) {
            //if we scroll to the last tab, alignment is set to the right side of the tab
            if (!this.chackActiveIndexRange()) return;
            const rectSide = this.props.activeIndex === this.props.children.length - 1 ? 'right' : 'left';
            this.scrollToIndex(this.props.activeIndex, rectSide);
        }
        // if prev state show arrow button, and current state doesn't show
        // need to reset the scroll position, or some tabs will be hided by container.
        if (prevState.showArrowButton && !this.state.showArrowButton) {
            this.scrollToZero();
        }

        if (prevProps.showModalButton !== this.props.showModalButton) {
            this.isShowModalButton();
        }

        if (prevProps.showArrowButton !== this.props.showArrowButton) {
            this.isShowArrowButton();
        }
        if (
            this.props.customStyle?.ActionButton &&
            prevProps.customStyle?.ActionButton !== this.props.customStyle?.ActionButton
        ) {
            this.FoldButton = makeFoldButton(this.props.customStyle?.ActionButton);
            this.ScrollButton = makeScrollButton(this.props.customStyle?.ActionButton);
        }
    }

    getTabNode(tab: HTMLDivElement & { __INTERNAL_NODE?: any }): React.ElementRef<'div'> {
        return tab.__INTERNAL_NODE;
    }

    unifyScrollMax(width: number) {
        return (width / 3) * 2;
    }

    handleScroll(direction: 'right' | 'left') {
        let leftMove = 0;
        const containerOffset = this.listContainer.getBoundingClientRect();
        const containerWidth = this.listContainer.offsetWidth;

        if (direction === 'right') {
            const tabLastOffset = this.getTabNode(this.tabRefs[this.tabRefs.length - 1]).getBoundingClientRect();

            leftMove = tabLastOffset.right - containerOffset.right;
            if (leftMove > containerWidth) {
                leftMove = this.unifyScrollMax(containerWidth);
            }
        } else if (direction === 'left') {
            const tabFirstOffset = this.getTabNode(this.tabRefs[0]).getBoundingClientRect();

            leftMove = tabFirstOffset.left - containerOffset.left;
            if (-leftMove > containerWidth) {
                leftMove = -this.unifyScrollMax(containerWidth);
            }
        }
        this.scrollPosition += leftMove;
        if (this.scrollPosition < 0) {
            this.scrollPosition = 0;
        }

        this.listScroll.style.transform = `translate3d(-${this.scrollPosition}px, 0, 0)`;
    }

    scrollToIndex(index: number, rectSide: 'left' | 'right') {
        const tabOffset = this.getTabNode(this.tabRefs[index]).getBoundingClientRect();
        const containerOffset = this.listContainer.getBoundingClientRect();
        console.log({ tabOffset, containerOffset });

        // Cancel scrolling if the tab is visible
        if (tabOffset.right < containerOffset.right && tabOffset.left > containerOffset.left) return;

        const leftMove = tabOffset['right'] + (rectSide === 'left' ? tabOffset['width'] : 0) - containerOffset['right'];
        console.log({ leftMove });

        this.scrollPosition += leftMove;
        if (this.scrollPosition < 0) {
            this.scrollPosition = 0;
        }
        console.log('this.scrollPosition', this.scrollPosition);

        this.listScroll.style.transform = `translate3d(-${this.scrollPosition}px, 0, 0)`;
    }

    scrollToZero() {
        this.listScroll.style.transform = `translate3d(0, 0, 0)`;
    }

    toggleModal() {
        this.setState({ modalIsOpen: !this.state.modalIsOpen });
    }

    isShowModalButton() {
        let { showModalButton } = this.props;
        if (isNumber(showModalButton)) {
            // $FlowFixMe, weired. currently set showModalButton as number | bool, but don't know why flow only can recognize it as bool
            showModalButton = this.props.children.length >= showModalButton;
        }
        this.setState({ showModalButton });
    }

    isShowArrowButton() {
        let { showArrowButton } = this.props;

        if (showArrowButton === 'auto') {
            let tabWidth = 0;
            const containerWidth = this.listContainer.offsetWidth;
            showArrowButton = false;
            for (let index = 0; index < this.tabRefs.length; index++) {
                const tab = this.getTabNode(this.tabRefs[index]);
                tabWidth += tab.offsetWidth;
                if (tabWidth >= containerWidth) {
                    showArrowButton = true;
                    break;
                }
            }
        }
        this.setState({ showArrowButton });
    }

    renderTabs(options: any = {}, isModal?: boolean) {
        const { children, activeIndex, handleTabChange, handleTabClose, customStyle } = this.props;
        const props = {
            handleTabChange,
            handleTabClose,
            CustomTabStyle: customStyle.Tab,
        };
        if (!isModal) {
            this.tabRefs = [];
        }
        return React.Children.map(children, (child, index) =>
            React.cloneElement(child as React.ReactElement, {
                key: index,
                active: index === activeIndex,
                index,
                tabIndex: index,
                ref: (node: HTMLDivElement) => {
                    if (!isModal && node) {
                        this.tabRefs.push(node);
                    }
                },
                ...props,
                ...options,
            })
        );
    }

    renderArrowButtons(ScrollButton: React.ElementType) {
        const { showArrowButton } = this.state;
        if (showArrowButton) {
            return (
                <div>
                    <ScrollButton
                        left
                        onClick={() => {
                            this.handleScroll('left');
                        }}
                        ref={(node: React.ReactElement) => (this.leftArrowNode = node)}
                        showModalButton={this.state.showModalButton}
                        className={'tabtab-arrow-button_left'}
                    >
                        <LeftIcon />
                    </ScrollButton>
                    <ScrollButton
                        onClick={() => {
                            this.handleScroll('right');
                        }}
                        ref={(node: React.ReactElement) => (this.rightArrowNode = node)}
                        className={'tabtab-arrow-button_right'}
                    >
                        <RightIcon />
                    </ScrollButton>
                </div>
            );
        }
        return null;
    }

    render() {
        const { ExtraButton } = this.props;
        const { modalIsOpen } = this.state;

        const TabList = this.TabList;
        const ScrollButton = this.ScrollButton;
        const FoldButton = this.FoldButton;

        invariant(this.props.children, 'React-tabtab Error: You MUST pass at least one tab');

        return (
            <div style={{ display: 'flex' }}>
                <TabList showModalButton={this.state.showModalButton} showArrowButton={this.state.showArrowButton}>
                    {this.state.showModalButton ? (
                        <FoldButton
                            ref={(node: React.ReactElement) => (this.foldNode = node)}
                            onClick={this.toggleModal}
                            showArrowButton={this.state.showArrowButton}
                        >
                            <BulletIcon />
                        </FoldButton>
                    ) : null}
                    {this.renderArrowButtons(ScrollButton)}
                    <ListInner ref={(node) => (this.listContainer = node)} className="tabtab-list-container">
                        <ListScroll ref={(node) => (this.listScroll = node)} role="tablist">
                            {this.renderTabs()}
                        </ListScroll>
                    </ListInner>
                </TabList>
                {!!ExtraButton && ExtraButton}
                {this.isShowModalButton && (
                    <TabListModal
                        isOpen={modalIsOpen}
                        onRequestClose={this.toggleModal}
                        dndContextProps={this.props.dndContextProps}
                        sortableContextProps={this.props.sortableContextProps}
                    >
                        {this.renderTabs({ vertical: true }, true)}
                    </TabListModal>
                )}
            </div>
        );
    }
}
