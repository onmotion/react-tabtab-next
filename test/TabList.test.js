import TabList from '../src/TabList';
import Tab from '../src/Tab';
import tabListTest from './tabListTest';

describe('TabList', () => {
  tabListTest('Tab', TabList, Tab);  
})
