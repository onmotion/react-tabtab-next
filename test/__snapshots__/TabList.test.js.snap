// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList custom ListStyle style 1`] = `
.c0 {
  background-color: white;
  text-align: left;
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  padding: 0 0px 0 0px;
}

.c1 {
  overflow: hidden;
}

.c2 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c3 {
  display: inline-block;
  padding: 10px 15px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3:hover {
  cursor: pointer;
  color: black;
}

.c4 {
  vertical-align: middle;
}

<TabList
  customStyle={
    Object {
      "CustomTabList": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": undefined,
        "componentStyle": ComponentStyle {
          "componentId": "sc-gZMcBi",
          "isStatic": false,
          "rules": Array [
            "
      background-color: white;
      text-align: left;
      position: relative;
      white-space: nowrap;
      overflow: hidden;
      width: auto;
      padding: ",
            [Function],
            ";
    ",
            "
          background-color: red;
        ",
          ],
        },
        "displayName": "Styled(styled.div)",
        "render": [Function],
        "styledComponentId": "sc-gZMcBi",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    }
  }
>
  <div>
    <styled.div
      hasExtraButton={false}
    >
      <StyledComponent
        forwardedClass={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": undefined,
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "c0",
              "rules": Array [
                "
            background-color: white;
            text-align: left;
            position: relative;
            white-space: nowrap;
            overflow: hidden;
            width: auto;
            padding: ",
                [Function],
                ";
          ",
              ],
            },
            "displayName": "styled.div",
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        hasExtraButton={false}
      >
        <div
          className="c0"
        >
          <styled.div>
            <StyledComponent
              forwardedClass={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": undefined,
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": true,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
                  overflow: hidden;
                ",
                    ],
                  },
                  "displayName": "styled.div",
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={[Function]}
            >
              <div
                className="c1"
              >
                <styled.ul
                  role="tablist"
                >
                  <StyledComponent
                    forwardedClass={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": undefined,
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-htpNat",
                          "isStatic": true,
                          "lastClassName": "c2",
                          "rules": Array [
                            "
                        padding-left: 0;
                        position: relative;
                        margin: 0;
                        list-style: none;
                        display: inline-block;
                        transition: transform .3s cubic-bezier(.42, 0, .58, 1);
                      ",
                          ],
                        },
                        "displayName": "styled.ul",
                        "render": [Function],
                        "styledComponentId": "sc-htpNat",
                        "target": "ul",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={[Function]}
                    role="tablist"
                  >
                    <ul
                      className="c2"
                      role="tablist"
                    >
                      <Tab
                        active={false}
                        index={0}
                        key="0/.0"
                        tabIndex={0}
                      >
                        <styled.li
                          active={false}
                          aria-controls="react-tabtab-panel-0"
                          aria-selected={false}
                          id="react-tabtab-tab-0"
                          onClick={[Function]}
                          role="tab"
                        >
                          <StyledComponent
                            active={false}
                            aria-controls="react-tabtab-panel-0"
                            aria-selected={false}
                            forwardedClass={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": undefined,
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-EHOje",
                                  "isStatic": false,
                                  "lastClassName": "c3",
                                  "rules": Array [
                                    "
                                display: ",
                                    [Function],
                                    ";
                                ",
                                    [Function],
                                    "
                              
                                user-select: none;
                                &:hover {
                                  cursor: pointer;
                                  color: black;
                                }
                              ",
                                  ],
                                },
                                "displayName": "styled.li",
                                "render": [Function],
                                "styledComponentId": "sc-EHOje",
                                "target": "li",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={[Function]}
                            id="react-tabtab-tab-0"
                            onClick={[Function]}
                            role="tab"
                          >
                            <li
                              aria-controls="react-tabtab-panel-0"
                              aria-selected={false}
                              className="c3"
                              id="react-tabtab-tab-0"
                              onClick={[Function]}
                              role="tab"
                            >
                              <styled.span>
                                <StyledComponent
                                  forwardedClass={
                                    Object {
                                      "$$typeof": Symbol(react.forward_ref),
                                      "attrs": undefined,
                                      "componentStyle": ComponentStyle {
                                        "componentId": "sc-bZQynM",
                                        "isStatic": true,
                                        "lastClassName": "c4",
                                        "rules": Array [
                                          "
                                      vertical-align: middle;
                                    ",
                                        ],
                                      },
                                      "displayName": "styled.span",
                                      "render": [Function],
                                      "styledComponentId": "sc-bZQynM",
                                      "target": "span",
                                      "toString": [Function],
                                      "warnTooManyClasses": [Function],
                                      "withComponent": [Function],
                                    }
                                  }
                                  forwardedRef={null}
                                >
                                  <span
                                    className="c4"
                                  >
                                    Tab1
                                  </span>
                                </StyledComponent>
                              </styled.span>
                            </li>
                          </StyledComponent>
                        </styled.li>
                      </Tab>
                    </ul>
                  </StyledComponent>
                </styled.ul>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </div>
</TabList>
`;

exports[`TabList render Tab List 1`] = `
.c0 {
  background-color: white;
  text-align: left;
  position: relative;
  white-space: nowrap;
  overflow: hidden;
  width: auto;
  padding: 0 0px 0 0px;
}

.c1 {
  overflow: hidden;
}

.c2 {
  padding-left: 0;
  position: relative;
  margin: 0;
  list-style: none;
  display: inline-block;
  -webkit-transition: -webkit-transform .3s cubic-bezier(.42,0,.58,1);
  -webkit-transition: transform .3s cubic-bezier(.42,0,.58,1);
  transition: transform .3s cubic-bezier(.42,0,.58,1);
}

.c3 {
  display: inline-block;
  padding: 10px 15px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c3:hover {
  cursor: pointer;
  color: black;
}

.c4 {
  vertical-align: middle;
}

<TabList
  customStyle={Object {}}
>
  <div>
    <styled.div
      hasExtraButton={false}
    >
      <StyledComponent
        forwardedClass={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": undefined,
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "c0",
              "rules": Array [
                "
            background-color: white;
            text-align: left;
            position: relative;
            white-space: nowrap;
            overflow: hidden;
            width: auto;
            padding: ",
                [Function],
                ";
          ",
              ],
            },
            "displayName": "styled.div",
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        hasExtraButton={false}
      >
        <div
          className="c0"
        >
          <styled.div>
            <StyledComponent
              forwardedClass={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": undefined,
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bwzfXH",
                    "isStatic": true,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
                  overflow: hidden;
                ",
                    ],
                  },
                  "displayName": "styled.div",
                  "render": [Function],
                  "styledComponentId": "sc-bwzfXH",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={[Function]}
            >
              <div
                className="c1"
              >
                <styled.ul
                  role="tablist"
                >
                  <StyledComponent
                    forwardedClass={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": undefined,
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-htpNat",
                          "isStatic": true,
                          "lastClassName": "c2",
                          "rules": Array [
                            "
                        padding-left: 0;
                        position: relative;
                        margin: 0;
                        list-style: none;
                        display: inline-block;
                        transition: transform .3s cubic-bezier(.42, 0, .58, 1);
                      ",
                          ],
                        },
                        "displayName": "styled.ul",
                        "render": [Function],
                        "styledComponentId": "sc-htpNat",
                        "target": "ul",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={[Function]}
                    role="tablist"
                  >
                    <ul
                      className="c2"
                      role="tablist"
                    >
                      <Tab
                        active={false}
                        index={0}
                        key="0/.0"
                        tabIndex={0}
                      >
                        <styled.li
                          active={false}
                          aria-controls="react-tabtab-panel-0"
                          aria-selected={false}
                          id="react-tabtab-tab-0"
                          onClick={[Function]}
                          role="tab"
                        >
                          <StyledComponent
                            active={false}
                            aria-controls="react-tabtab-panel-0"
                            aria-selected={false}
                            forwardedClass={
                              Object {
                                "$$typeof": Symbol(react.forward_ref),
                                "attrs": undefined,
                                "componentStyle": ComponentStyle {
                                  "componentId": "sc-EHOje",
                                  "isStatic": false,
                                  "lastClassName": "c3",
                                  "rules": Array [
                                    "
                                display: ",
                                    [Function],
                                    ";
                                ",
                                    [Function],
                                    "
                              
                                user-select: none;
                                &:hover {
                                  cursor: pointer;
                                  color: black;
                                }
                              ",
                                  ],
                                },
                                "displayName": "styled.li",
                                "render": [Function],
                                "styledComponentId": "sc-EHOje",
                                "target": "li",
                                "toString": [Function],
                                "warnTooManyClasses": [Function],
                                "withComponent": [Function],
                              }
                            }
                            forwardedRef={[Function]}
                            id="react-tabtab-tab-0"
                            onClick={[Function]}
                            role="tab"
                          >
                            <li
                              aria-controls="react-tabtab-panel-0"
                              aria-selected={false}
                              className="c3"
                              id="react-tabtab-tab-0"
                              onClick={[Function]}
                              role="tab"
                            >
                              <styled.span>
                                <StyledComponent
                                  forwardedClass={
                                    Object {
                                      "$$typeof": Symbol(react.forward_ref),
                                      "attrs": undefined,
                                      "componentStyle": ComponentStyle {
                                        "componentId": "sc-bZQynM",
                                        "isStatic": true,
                                        "lastClassName": "c4",
                                        "rules": Array [
                                          "
                                      vertical-align: middle;
                                    ",
                                        ],
                                      },
                                      "displayName": "styled.span",
                                      "render": [Function],
                                      "styledComponentId": "sc-bZQynM",
                                      "target": "span",
                                      "toString": [Function],
                                      "warnTooManyClasses": [Function],
                                      "withComponent": [Function],
                                    }
                                  }
                                  forwardedRef={null}
                                >
                                  <span
                                    className="c4"
                                  >
                                    Tab1
                                  </span>
                                </StyledComponent>
                              </styled.span>
                            </li>
                          </StyledComponent>
                        </styled.li>
                      </Tab>
                    </ul>
                  </StyledComponent>
                </styled.ul>
              </div>
            </StyledComponent>
          </styled.div>
        </div>
      </StyledComponent>
    </styled.div>
  </div>
</TabList>
`;
